internal/iam/testing.go

```diff
@@ -345,6 +315,148 @@ func TestManagedGroupRole(t testing.TB, conn *db.DB, roleId, managedGrpId string
 	return r
 }
 
+// TestRoleWithGrants creates a role suitable for testing along with grants
+// Functional options for GrantScopes aren't used to express that
+// this function does not provide any default grant scope unlike TestRole
+func TestRoleWithGrants(t testing.TB, conn *db.DB, scopeId string, grantScopeIDs []string, grants []string) *Role {
+	t.Helper()
+
+	ctx := context.Background()
+	require := require.New(t)
+	rw := db.New(conn)
+
+	role, err := NewRole(ctx, scopeId)
+	require.NoError(err)
+	id, err := newRoleId(ctx)
+	require.NoError(err)
+	role.PublicId = id
+	require.NoError(rw.Create(ctx, role))
+	require.NotEmpty(role.PublicId)
+
+	for _, gsi := range grantScopeIDs {
+		gs, err := NewRoleGrantScope(ctx, id, gsi)
+		require.NoError(err)
+		require.NoError(rw.Create(ctx, gs))
+		role.GrantScopes = append(role.GrantScopes, gs)
+	}
+	for _, g := range grants {
+		_ = TestRoleGrant(t, conn, role.PublicId, g)
+	}
+	return role
+}
+
+type TestRoleGrantsRequest struct {
+	RoleScopeID string
+	GrantScopes []string
+	Grants      []string
+}
+
+// TestUserManagedGroupGrantsFunc returns a function that creates a user which has been given
+// the request grants through managed group.
+// Note: This method is not responsible for associating the user to the managed group. That action needs to be done
+// by the caller
+// This function returns iam.User and the AccountID from the account setup func
+func TestUserManagedGroupGrantsFunc(
+	t *testing.T,
+	conn *db.DB,
+	kmsCache *kms.Kms,
+	scopeID string,
+	managedGroupAccountSetupFunc func() (string, string),
```

Comment:
I think we could define some types for some of these functions to get stricter type checking. Let's sync up since it would involve several changes in various places, so it isn't very easy to show in a PR comment.